README.md
# create-svelte

Everything you need to build a Svelte project, powered by [`create-svelte`](https://github.com/sveltejs/kit/tree/master/packages/create-svelte);

## Creating a project

If you're seeing this, you've probably already done this step. Congrats!

```bash
# create a new project in the current directory
npm init svelte@next

# create a new project in my-app
npm init svelte@next my-app
```

> Note: the `@next` is temporary

## Developing

Once you've created a project and installed dependencies with `npm install` (or `pnpm install` or `yarn`), start a development server:

```bash
npm run dev

# or start the server and open the app in a new browser tab
npm run dev -- --open
```

## Building

Before creating a production version of your app, install an [adapter](https://kit.svelte.dev/docs#adapters) for your target environment. Then:

```bash
npm run build
```

> You can preview the built app with `npm run preview`, regardless of whether you installed an adapter. This should _not_ be used to serve your app in production.

## Updating

To update the repo, do the following:
1. Ensure that the main gcsim repo is up-to-date with any desired changes
2. In the directory root, run `scripts/bump.sh "[TYPE]"`, where `[TYPE]` can be any of `major`, `minor`, or `patch` (usually you want patch)
3. Update the `release-notes.txt` file appropriately with updates from the main repo
4. Push to the repo, which triggers the Github workflow to rebuild and redeploy


release-notes.txt
0.4.37 - Feb 08, 2022
---------------------

- Fixed Zhongli Stele not snapshotting
- Fixed Zhongli Stele max count incorrectly capped at 1
- Fixed Xiao snapshot timing
- Fixed Yunjin perfect counter. Changed to skill[perfect=1]
- Fixed Kazuha c2 not applied to self when off-field


0.4.36 - Feb 06, 2022
---------------------

- Fixes pale flame 2 pc set typo
- Fixes Bennett burst frame timing and A4 activation method
- Fixes Gorou C6 incorrectly snapshotting
- Add max energy in energy log
- Add Eula skill hold check for tenacity and pale flame
- Fixed Eula burst timing
- Fixed Skyward Pride proc count
- Added Calamity Queller
- Fix Polar Star
- Fix Mistsplitter burst bug
- Add Moonglow


0.4.35 - Feb 01, 2022
---------------------

- Fix Aloy coil stacks issue
- Fix Xingqiu infinite burst
- Fix Venti Q frames
- Fix Kokomi not adding in heal bonus
- Add ability to set Geo Traveller burst construct count for Zhongli
- Fix Keqing C4 and weapon attack tag checks
- Add Polar Star

0.4.34 - Jan 27, 2022
---------------------

- Overall debug improvements, moving things into consistent categories (#201)
- A lot of bugfixes. Below list the ones that have been resolved:
  - Bennett Q did not work properly at C0
  - Lisa C1 was activating on her CA and not her hold E
  - Windblume Ode did not have an onfield check
  - Ganyu E explosion particle timing was wrong
  - Kaeya CA was coded as being an NA tag
  - Geo Traveller had an 80 cost burst instead of 60
  - Geo Traveller energy refund did not work if cast at frame 0
  - Hydro swirl was doing AoE damage
  - Jean NA was snapshotting incorrectly
  - Jean CA was not implemented
  - Keqing Q was coded as snapshotted, not dynamic

0.4.33 - Jan 24, 2022
---------------------

- Fix substat optimizer typo
- Fix Amber Attack Tag, ICD Tag, and ICD Group
- Fix Eula grimheart stacks not adding
- Fix Fischl snapshot frames to account for on damage proc timing
- Fix Freedom Sworn - ATK did not snapshot, dmg% buff did not reset

0.4.32 - Jan 24, 2022
---------------------

- Implement substat optimization in cmd line
- Fix Primordial Jade Winged-Spear stacks
- Fix Guoba snapshot timing
- Fix Kazuha plunge not shattering
- Fix Chongyun infuse field timing
- Fix swirl damage ICD so an enemy can be affected by more than 2 damage instances
- Fix VV not proccing on AoE targets appropriately

0.4.31 - Jan 22, 2022
---------------------

- Fix Rosaria burst using wrong multiplier
- Fix Diona burst framess

0.4.30 - Jan 21, 2022
---------------------

- Fix Albedo burst CD
- Fix Barbara E scaling

0.4.29 - Jan 20, 2022
---------------------

- Fix burst CD start to match energy consumption
- Fix Ayaka ICD on NA and Dash

0.4.28 - Jan 18, 2022
---------------------

- Fix Chongyun Q frames

0.4.27 - Jan 15, 2022
---------------------

- Minor bug fixes

0.4.26 - Jan 11, 2022
---------------------

- Bug fixes for Ningguang, prototype amber, and elegy
- Disable debug mode option; always use debug mode
- Fix missing images

0.4.25 - Jan 11, 2022
---------------------

- Logging cleanup

0.4.24 - Jan 08, 2022
---------------------

- Fixed Solar Pearl, Alley Hunter, Lost Prayer, Lisa bugs

0.4.23 - Jan 08, 2022
---------------------

- Fixed Ganyu Q bug

0.4.22 - Jan 08, 2022
---------------------

- Fixed output file name

0.4.21 - Jan 08, 2022
---------------------

- Disable workflow cache

0.4.20 - Jan 07, 2022
---------------------

- Shenhe ICD fix
- Yunjin particles adjustment
- Added share button


0.4.19 - Jan 07, 2022
---------------------

- Actually fix shatter this time.. (forgot to merge bugfix)

0.4.18 - Jan 07, 2022
---------------------

- Fix shatter bug
- Enable mods by default in debug view

0.4.17 - Jan 07, 2022
---------------------

- Forgot to enable Yunjin...

0.4.16 - Jan 06, 2022
---------------------

- Fix Shenhe NO4 snapshotting
- Fix Kazuha frames
- Fix shields not overwriting properly

0.4.15 - Jan 06, 2022
---------------------

- Actually fix Shenhe cooldown this time...
- Additional bug fixes for Noelle,  Hutao, Bennett, Yoimiya


0.4.14 - Jan 05, 2022
---------------------

- Fix Shenhe skill cd


0.4.13 - Jan 05, 2022
---------------------

- Add Shenhe
- Fixed various Zhongli bugs
- Change +try to either +try=drop or +try=wait

0.4.12 - Jan 02, 2022
---------------------

- Various bug fixes

0.4.11 - Jan 02, 2022
---------------------

- Fixed bug introduced in 0.4.10

0.4.10 - Jan 02, 2022
---------------------

- Fixed Kazuha burst timing
- Fixed Beidou ascension stat
- Fixed debug output for reaction into freeze/ec

0.4.9 - Jan 02, 2022
--------------------

- Added guoba swirl
- Fixed Noelle and HT snapshots
- Some backend clean up
- Enabled Gorou

0.4.8 - Dec 31, 2021
--------------------

- Fix limit/reset_limit commands not work

0.4.7 - Dec 31, 2021
--------------------

- Add restart command to calc mode
- Various bug fixes

0.4.6 - Dec 31, 2021
--------------------

- Add Bennett Q self-infusion

0.4.5 - Dec 31, 2021
--------------------

- Revert changes made in 0.4.4

0.4.4 - Dec 30, 2021
--------------------

- Simplified calc mode shorthand to remove commas

0.4.3 - Dec 30, 2021
--------------------

- Allow shorthand in calc mode (ui only)

0.4.2 - Dec 30, 2021
--------------------

- Syntax bug fixes

0.4.1 - Dec 30, 2021
--------------------

- Fixed ICD group bug

0.4.0 - Dec 29, 2021
--------------------

- THIS RELEASE CONTAINS BREAKING CHANGES!!!
  - The entire action list syntax has been redesigned to be simpler.
  - All action lists on https://www.gcsim.app have been updated to this new format. Please check for example
  - Documentation will be added shortly to the wiki


src/util/good.ts
export interface IGOOD {
  format: "GOOD"; //A way for people to recognize this format.
  version: number; //API version.
  source: string; //the app that generates this data.
  characters?: ICharacter[];
  artifacts?: IArtifact[];
  weapons?: IWeapon[];
}

export interface IArtifact {
  setKey: SetKey; //e.g. "GladiatorsFinale"
  slotKey: SlotKey; //e.g. "plume"
  level: number; //0-20 inclusive
  rarity: number; //1-5 inclusive
  mainStatKey: StatKey;
  location: CharacterKey | ""; //where "" means not equipped.
  lock: boolean; //Whether the artifact is locked in game.
  substats: ISubstat[];
}

export interface ISubstat {
  key: StatKey; //e.g. "critDMG_"
  value: number; //e.g. 19.4
}

export type SlotKey = "flower" | "plume" | "sands" | "goblet" | "circlet";

export interface IWeapon {
  key: WeaponKey; //"CrescentPike"
  level: number; //1-90 inclusive
  ascension: number; //0-6 inclusive. need to disambiguate 80/90 or 80/80
  refinement: number; //1-5 inclusive
  location: CharacterKey | ""; //where "" means not equipped.
  lock: boolean; //Whether the weapon is locked in game.
}

export interface ICharacter {
  key: CharacterKey; //e.g. "Rosaria"
  level: number; //1-90 inclusive
  constellation: number; //0-6 inclusive
  ascension: number; //0-6 inclusive. need to disambiguate 80/90 or 80/80
  talent: {
    //does not include boost from constellations. 1-15 inclusive
    auto: number;
    skill: number;
    burst: number;
  };
}

export const allElements = [
  "anemo",
  "geo",
  "electro",
  "hydro",
  "pyro",
  "cryo",
] as const;
export type ElementKey = typeof allElements[number];

export type StatKey =
  | "hp" //HP
  | "hp_" //HP%
  | "atk" //ATK
  | "atk_" //ATK%
  | "def" //DEF
  | "def_" //DEF%
  | "eleMas" //Elemental Mastery
  | "enerRech_" //Energy Recharge%
  | "heal_" //Healing Bonus%
  | "critRate_" //CRIT Rate%
  | "critDMG_" //CRIT DMG%
  | "physical_dmg_" //Physical DMG Bonus%
  | "anemo_dmg_" //Anemo DMG Bonus%
  | "geo_dmg_" //Geo DMG Bonus%
  | "electro_dmg_" //Electro DMG Bonus%
  | "hydro_dmg_" //Hydro DMG Bonus%
  | "pyro_dmg_" //Pyro DMG Bonus%
  | "cryo_dmg_"; //Cryo DMG Bonus%

type SetKey =
  | "Adventurer" //Adventurer
  | "ArchaicPetra" //Archaic Petra
  | "Berserker" //Berserker
  | "BlizzardStrayer" //Blizzard Strayer
  | "BloodstainedChivalry" //Bloodstained Chivalry
  | "BraveHeart" //Brave Heart
  | "CrimsonWitchOfFlames" //Crimson Witch of Flames
  | "DefendersWill" //Defender's Will
  | "EmblemOfSeveredFate" //Emblem of Severed Fate
  | "Gambler" //Gambler
  | "GladiatorsFinale" //Gladiator's Finale
  | "HeartOfDepth" //Heart of Depth
  | "Instructor" //Instructor
  | "Lavawalker" //Lavawalker
  | "LuckyDog" //Lucky Dog
  | "MaidenBeloved" //Maiden Beloved
  | "MartialArtist" //Martial Artist
  | "NoblesseOblige" //Noblesse Oblige
  | "PaleFlame" //Pale Flame
  | "PrayersForDestiny" //Prayers for Destiny
  | "PrayersForIllumination" //Prayers for Illumination
  | "PrayersForWisdom" //Prayers for Wisdom
  | "PrayersToSpringtime" //Prayers to Springtime
  | "ResolutionOfSojourner" //Resolution of Sojourner
  | "RetracingBolide" //Retracing Bolide
  | "Scholar" //Scholar
  | "ShimenawasReminiscence" //Shimenawa's Reminiscence
  | "TenacityOfTheMillelith" //Tenacity of the Millelith
  | "TheExile" //The Exile
  | "ThunderingFury" //Thundering Fury
  | "Thundersoother" //Thundersoother
  | "TinyMiracle" //Tiny Miracle
  | "TravelingDoctor" //Traveling Doctor
  | "ViridescentVenerer" //Viridescent Venerer
  | "WanderersTroupe"; //Wanderer's Troupe

type CharacterKey =
  | "Albedo" //Albedo
  | "Aloy" //Aloy
  | "Amber" //Amber
  | "Barbara" //Barbara
  | "Beidou" //Beidou
  | "Bennett" //Bennett
  | "Chongyun" //Chongyun
  | "Diluc" //Diluc
  | "Diona" //Diona
  | "Eula" //Eula
  | "Fischl" //Fischl
  | "Ganyu" //Ganyu
  | "HuTao" //Hu Tao
  | "Jean" //Jean
  | "KaedeharaKazuha" //Kaedehara Kazuha
  | "Kaeya" //Kaeya
  | "KamisatoAyaka" //Kamisato Ayaka
  | "Keqing" //Keqing
  | "Klee" //Klee
  | "KujouSara" //Kujou Sara
  | "Lisa" //Lisa
  | "Mona" //Mona
  | "Ningguang" //Ningguang
  | "Noelle" //Noelle
  | "Qiqi" //Qiqi
  | "RaidenShogun" //Raiden Shogun
  | "Razor" //Razor
  | "Rosaria" //Rosaria
  | "SangonomiyaKokomi" //Sangonomiya Kokomi
  | "Sayu" //Sayu
  | "Sucrose" //Sucrose
  | "Tartaglia" //Tartaglia
  | "Traveler" //Traveler
  | "Venti" //Venti
  | "Xiangling" //Xiangling
  | "Xiao" //Xiao
  | "Xingqiu" //Xingqiu
  | "Xinyan" //Xinyan
  | "Yanfei" //Yanfei
  | "Yoimiya" //Yoimiya
  | "Zhongli"; //Zhongli

type WeaponKey =
  | "Akuoumaru" //Akuoumaru
  | "AlleyHunter" //Alley Hunter
  | "AmenomaKageuchi" //Amenoma Kageuchi
  | "AmosBow" //Amos' Bow
  | "ApprenticesNotes" //Apprentice's Notes
  | "AquilaFavonia" //Aquila Favonia
  | "BeginnersProtector" //Beginner's Protector
  | "BlackTassel" //Black Tassel
  | "BlackcliffAgate" //Blackcliff Agate
  | "BlackcliffLongsword" //Blackcliff Longsword
  | "BlackcliffPole" //Blackcliff Pole
  | "BlackcliffSlasher" //Blackcliff Slasher
  | "BlackcliffWarbow" //Blackcliff Warbow
  | "BloodtaintedGreatsword" //Bloodtainted Greatsword
  | "CompoundBow" //Compound Bow
  | "CoolSteel" //Cool Steel
  | "CrescentPike" //Crescent Pike
  | "DarkIronSword" //Dark Iron Sword
  | "Deathmatch" //Deathmatch
  | "DebateClub" //Debate Club
  | "DodocoTales" //Dodoco Tales
  | "DragonsBane" //Dragon's Bane
  | "DragonspineSpear" //Dragonspine Spear
  | "DullBlade" //Dull Blade
  | "ElegyForTheEnd" //Elegy for the End
  | "EmeraldOrb" //Emerald Orb
  | "EngulfingLightning" //Engulfing Lightning
  | "EverlastingMoonglow" //Everlasting Moonglow
  | "EyeOfPerception" //Eye of Perception
  | "FavoniusCodex" //Favonius Codex
  | "FavoniusGreatsword" //Favonius Greatsword
  | "FavoniusLance" //Favonius Lance
  | "FavoniusSword" //Favonius Sword
  | "FavoniusWarbow" //Favonius Warbow
  | "FerrousShadow" //Ferrous Shadow
  | "FesteringDesire" //Festering Desire
  | "FilletBlade" //Fillet Blade
  | "FreedomSworn" //Freedom-Sworn
  | "Frostbearer" //Frostbearer
  | "HakushinRing" //Hakushin Ring
  | "Halberd" //Halberd
  | "Hamayumi" //Hamayumi
  | "HarbingerOfDawn" //Harbinger of Dawn
  | "HuntersBow" //Hunter's Bow
  | "IronPoint" //Iron Point
  | "IronSting" //Iron Sting
  | "KatsuragikiriNagamasa" //Katsuragikiri Nagamasa
  | "KitainCrossSpear" //Kitain Cross Spear
  | "LionsRoar" //Lion's Roar
  | "LithicBlade" //Lithic Blade
  | "LithicSpear" //Lithic Spear
  | "LostPrayerToTheSacredWinds" //Lost Prayer to the Sacred Winds
  | "LuxuriousSeaLord" //Luxurious Sea-Lord
  | "MagicGuide" //Magic Guide
  | "MappaMare" //Mappa Mare
  | "MemoryOfDust" //Memory of Dust
  | "Messenger" //Messenger
  | "MistsplitterReforged" //Mistsplitter Reforged
  | "MitternachtsWaltz" //Mitternachts Waltz
  | "OldMercsPal" //Old Merc's Pal
  | "OtherworldlyStory" //Otherworldly Story
  | "PocketGrimoire" //Pocket Grimoire
  | "PolarStar" //Polar Star
  | "Predator" //Predator
  | "PrimordialJadeCutter" //Primordial Jade Cutter
  | "PrimordialJadeWingedSpear" //Primordial Jade Winged-Spear
  | "PrototypeAmber" //Prototype Amber
  | "PrototypeArchaic" //Prototype Archaic
  | "PrototypeCrescent" //Prototype Crescent
  | "PrototypeRancour" //Prototype Rancour
  | "PrototypeStarglitter" //Prototype Starglitter
  | "Rainslasher" //Rainslasher
  | "RavenBow" //Raven Bow
  | "RecurveBow" //Recurve Bow
  | "RoyalBow" //Royal Bow
  | "RoyalGreatsword" //Royal Greatsword
  | "RoyalGrimoire" //Royal Grimoire
  | "RoyalLongsword" //Royal Longsword
  | "RoyalSpear" //Royal Spear
  | "Rust" //Rust
  | "SacrificialBow" //Sacrificial Bow
  | "SacrificialFragments" //Sacrificial Fragments
  | "SacrificialGreatsword" //Sacrificial Greatsword
  | "SacrificialSword" //Sacrificial Sword
  | "SeasonedHuntersBow" //Seasoned Hunter's Bow
  | "SerpentSpine" //Serpent Spine
  | "SharpshootersOath" //Sharpshooter's Oath
  | "SilverSword" //Silver Sword
  | "SkyriderGreatsword" //Skyrider Greatsword
  | "SkyriderSword" //Skyrider Sword
  | "SkywardAtlas" //Skyward Atlas
  | "SkywardBlade" //Skyward Blade
  | "SkywardHarp" //Skyward Harp
  | "SkywardPride" //Skyward Pride
  | "SkywardSpine" //Skyward Spine
  | "Slingshot" //Slingshot
  | "SnowTombedStarsilver" //Snow-Tombed Starsilver
  | "SolarPearl" //Solar Pearl
  | "SongOfBrokenPines" //Song of Broken Pines
  | "StaffOfHoma" //Staff of Homa
  | "SummitShaper" //Summit Shaper
  | "SwordOfDescension" //Sword of Descension
  | "TheAlleyFlash" //The Alley Flash
  | "TheBell" //The Bell
  | "TheBlackSword" //The Black Sword
  | "TheCatch" //"The Catch"
  | "TheFlute" //The Flute
  | "TheStringless" //The Stringless
  | "TheUnforged" //The Unforged
  | "TheViridescentHunt" //The Viridescent Hunt
  | "TheWidsith" //The Widsith
  | "ThrillingTalesOfDragonSlayers" //Thrilling Tales of Dragon Slayers
  | "ThunderingPulse" //Thundering Pulse
  | "TravelersHandySword" //Traveler's Handy Sword
  | "TwinNephrite" //Twin Nephrite
  | "VortexVanquisher" //Vortex Vanquisher
  | "WasterGreatsword" //Waster Greatsword
  | "WhiteIronGreatsword" //White Iron Greatsword
  | "WhiteTassel" //White Tassel
  | "Whiteblind" //Whiteblind
  | "WindblumeOde" //Windblume Ode
  | "WineAndSong" //Wine and Song
  | "WolfsGravestone"; //Wolf's Gravestone
